Allow TaskGroup's ChildTaskResult Type To Be Inferred

rlziii opened the pull request:
Original discussion here: https://forums.swift.org/t/allow-taskgroups-childtaskresult-type-to-be-inferred/72175

See: https://github.com/apple/swift/pull/74517

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,138 @@
+
+## Introduction
+
+`TaskGroup` and `ThrowingTaskGroup` currently require that one of their two generics (`ChildTaskResult`) always be specified upon creation; this can be simplified by allowing the compiler to infer both of the generics in most cases.
```
```suggestion
`TaskGroup` and `ThrowingTaskGroup` currently require that one of their two generics (`ChildTaskResult`) always be specified upon creation. Due to improvements in closure parameter/result type inference introduced by [SE-0326](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0326-extending-multi-statement-closure-inference.md) this can be simplified by allowing the compiler to infer both of the generics in most cases.
```

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
````diff
@@ -0,0 +1,138 @@
+}
+```
+
+The type of `s` (which is the `GroupResult` generic) is inferred above (as `String`).  However, the return value of the `addTask` closures cannot be inferred and must be supplied via `of childTaskResultType: ChildTaskResult.Type` (above as `Void`).
````
```suggestion
The type of `s` (which is the `GroupResult` type) is correctly inferred as `String`.  However, the return value of the `addTask` closures cannot is not inferred and currently must be supplied to the `of:`parameter of the withTaskGroup function (e.g. `Void`).
```

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,138 @@
+
+## ABI compatibility
+
+No ABI impact since this is an additive change to default arguments.
```
```suggestion
No ABI impact since adding a default argument value is binary compatible change.
```

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,138 @@
+
+## Future directions
+
+...
```
Maybe let's add this:

---


### TaskGroup APIs without the "with..." closures

While not possible without more compiler features to enforce the safety of a task group not escaping a context, and having to await all of its results at the end of a scope... it is an interesting future direction to explore a TaskGroup API that does not need to resort to `with...` methods, like this:

```
// Potential long-term direction that might be possible
func test() async { 
  let group = TaskGroup<Int>()
  group.addTask { ... }
  
  // going out of scope would have to imply group.awaitAll()
}
```

If we were to explore such API, the type inference rules would be somewhat different, and a TaskGroup would likely be initialized more similarly to a collection: `TaskGroup<Int>`.

This proposal has no impact on this future direction, and can be accepted as is, without precluding future developments in API ergonomics like this.

ktoso reviewed (approved):
Looks good to me! Minor ideas to improve proposal and we can ask the core team for a review 👍

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,138 @@
+
+## Acknowledgments
+
+...
```
If you'd like to drop a small nod here for review/confirming viability to myself and @xedin that'd be nice, your call though :)

ktoso commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,138 @@
+* Author: [Richard L Zarth III](https://github.com/rlziii)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN)
```
```suggestion
* Implementation: [apple/swift#74517](https://github.com/apple/swift/pull/74517)
```

rlziii commented:
> Looks good to me! Minor ideas to improve proposal and we can ask the core team for a review 👍

Thank you for looking over everything.  I have implemented the suggestions and updated a spots of grammar.  Should I go ahead and tag this as `Ready for review`?

ktoso commented:
Yeap, please mark as ready to review! :)

DougGregor commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,154 @@
+// Expect `ChildTaskResult` to be `Int`...
+withTaskGroup(of: Int.self) { group in
+    group.cancelAll()
+    group.addTask { 42 }
```
I think it would be beneficial to have some more realistic examples. For example, there are small-but-realistic examples at https://www.avanderlee.com/concurrency/task-groups-in-swift/ and https://www.kodeco.com/books/modern-concurrency-in-swift/v1.0/chapters/7-concurrent-code-with-taskgroup that you could reference (with attribution!) and will give reviewers a better feel for how this impacts use of the feature.

rlziii replied:
I have updated the examples to use more real-world scenarios.  Thank you for the suggestion.  👍

DougGregor commented on proposals/NNNN-allow-taskgroup-childtaskresult-type-to-be-inferred.md:
```diff
@@ -0,0 +1,154 @@
+
+## Detailed design
+
+Because type inference is top-down, it relies on the first statement that uses `group` to infer determine the generic for `ChildTaskResult`.  Therefore, it is possible to get a compiler error by creating a task group where the first use of `group` does not use `addTask(...)`, like so:
```
Thank you! I'll kick the review off shortly.
